### PR TITLE
Issue/241

### LINKED ISSUE #241: Update image tag
To make creating/editing/translating images more user friendly, the following change needs to be implemented:
Currently an image looks like this:
```
{image file-path="http://placehold.it/350x150" caption="Description of the image."}
```
And the suggested change is:
```
{image file-path="http://placehold.it/350x150" caption="true"}

Description of the image.

{image end}
```
Where the caption is the text between the blocks only if `caption` is `true`. There are multiple other parameters that the image tag uses (e.g. alt, alignment, hover-text, etc) but we have decided to pull the caption out as this the most obvious user facing text and therefore want to make it easy for editors to make changes.

This is only the suggested solution and is up for discussion (e.g. do we need to specify `caption="true"`? what happens when there is no caption, do we still use `{image end}` or make it an inline processor?) cc @JackMorganNZ @jordangriffiths01

### PR DESCRIPTION
**Pull request by hayleyavw**
Fixes #241 

Complete with docs and tests.

To enable parsing of the two different kinds of blocks, depending on whether a caption is given or not (see description in issue #241), the image processor was removed and replaced with and `ImageTagBlockProcessor` and `ImageContainerBlockProcessor`, each acting as subclasses for the corresponding generic processor.

### CODE REVIEW

File: verto/processors/GenericContainerBlockProcessor.py [hunk 1]
```diff
@@ -113,6 +115,7 @@ def run(self, parent, blocks):
             raise ArgumentValueError(self.processor, 'content', content, message)
 
         argument_values['content'] = content
+# think there is something wrong here
```
Comment (jordangriffiths01) [inline verto/processors/GenericContainerBlockProcessor.py]: Do you still think something is wrong here? :joy:

File: verto/tests/ConfigurationTest.py [hunk 2]
```diff
@@ -141,8 +140,7 @@ def test_multiple_calls(self):
             verto.clear_saved_data()
 
     def test_multiple_calls_without_clearing(self):
-        '''Tests that if the verto extension is not cleared that
-        information such as required_files and slugs are persistant.
+        '''Tests that if the verto extension is not cleared that information such as required_files and slugs are persistant.
```
Comment (jordangriffiths01) [inline verto/tests/ConfigurationTest.py]: s/persistant/persistent

File: verto/tests/ImageTagTest.py [hunk 3]
```diff
@@ -321,16 +303,17 @@ def test_image_in_numbered_list(self):
         expected_images = set()
         self.assertSetEqual(expected_images, images)
 
-    #~
-    # Doc Tests
-    #~
+    # #~
```
Comment (jordangriffiths01) [inline verto/tests/ImageTagTest.py]: Reason for double comment here?

File: verto/processors/GenericContainerBlockProcessor.py (hunk 1 above)
Comment (hayleyavw) [reply verto/processors/GenericContainerBlockProcessor.py]: lol

Comment (hayleyavw) [reply verto/processors/GenericContainerBlockProcessor.py]: (also, fixed, just not shown as outdated)

File: docs/source/contributing.rst [hunk 4]
```diff
@@ -152,6 +152,7 @@ The json description of a generic processor must contain the attributes:
   - ``arguments``: An object describing arguments passed to the tag.
   - ``template_parameters``: An object describing template parameters.
   - (Optional) ``template_name``: A custom name for the html-template to use. Defaults to the processor name otherwise.
+  - (Optional) ``tag_argument``: The text given at the beginning of a tag (e.g. the tag argument for ``{image file-path="example.png"}`` is ``image``). This is only necessary for processors with different names sharing the same resources (e.g. both ``image-container`` and ``image-tag`` share the ``image.html`` template).
```
Comment (JackMorganNZ) [inline docs/source/contributing.rst]: Confused why this talks about tag label then goes on about HTML templates?

File: verto/errors/ImageCaptionContainsImageError.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
+from verto.errors.Error import Error
+
+
+class ImageCaptionContainsImageError(Error):
+    '''Exception raised when a caption includes an image block.
+
+    Attributes:
+        tag: tag which was not matched
+        block:
```
Comment (JackMorganNZ) [inline verto/errors/ImageCaptionContainsImageError.py]: Are these supposed to be blank?

File: verto/errors/ImageMissingCaptionError.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
+from verto.errors.Error import Error
+
+
+class ImageMissingCaptionError(Error):
+    '''Exception raised when an image is missing a subtitle.
+
+    Attributes:
+        tag: tag which was not matched
+        block: block where tag was not matched
```
Comment (JackMorganNZ) [inline verto/errors/ImageMissingCaptionError.py]: This isn't used in this module.

File: docs/source/contributing.rst (hunk 4 above)
Comment (hayleyavw) [reply docs/source/contributing.rst]: An image how has multiple processors introduced in this PR, `image-container` and `image-tag`, but both of these processors are used in the markdown files like so:
```
{image file-path="an-image.png"}
```
Therefore I introduced a new optional attribute to the processor definitions in the JSON file to specify the opening tag argument for a processor, the html part is simply giving an example of why you would do this.

Comment (JackMorganNZ) [reply docs/source/contributing.rst]: Is `tag_argument` tied to the HTML template that is used? I'm confused to why it doesn't say:

> (e.g. both ``image-container`` and ``image-tag`` share the ``tag_argument``).

Comment (hayleyavw) [reply docs/source/contributing.rst]: OH I see your point, will change to your suggestion for clarification :+1:

**Pull request merged**